test/cpp/api/CMakeLists.txt

```diff
@@ -65,6 +66,14 @@ if(NOT MSVC)
   target_compile_options_if_supported(test_api "-Wno-maybe-uninitialized")
   # gcc gives nonsensical warnings about variadic.h
   target_compile_options_if_supported(test_api "-Wno-unused-but-set-parameter")
+  
+  # Add -Wno-error=nonnull for GCC 12+
+  if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 12)
+    check_cxx_compiler_flag("-Wno-error=nonnull" HAS_WNO_ERROR_NONNULL)
+    if(HAS_WNO_ERROR_NONNULL)
+      target_compile_options(test_api PRIVATE "-Wno-error=nonnull")
+    endif()
+  endif()
 endif()
```

Comment:
Can use 
```
 target_compile_options_if_supported("-Wno-error=nonnull") 
```
?

Reply:
Hi @cyyever 

As the issue comes only with gcc compiler and that too with gcc >=12. So, i added check for GCC compiler only.
And also, I want to ensure that the flag `-Wno-error=nonnull` is added specifically to the test_api target, not globally

Reply:
I include include(CheckCXXCompilerFlag) as i also need to ensure whether a compiler in all environment support 
` -Wno-error=nonnull` by using `check_cxx_compiler_flag`

if support then only add ` -Wno-error=nonnull`

Reply:
Hi @cyyever 
As per your suggestion I guess below code block will be good to add only.
Correct me if I am wrong
```
# Add -Wno-error=nonnull for GCC 12+  
if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 12) 
    target_compile_options_if_supported(test_api PRIVATE "-Wno-error=nonnull") 
endif()
```

Reply:
It is worth trying

Reply:
```
# Add -Wno-error=nonnull for GCC 12+  
if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 12) 
    target_compile_options_if_supported(test_api PRIVATE "-Wno-error=nonnull") 
endif()
```
If I only add the above code, I face a similar issue.
After removing PRIVATE, then only build is successful.

But removing PRIVATE does not ensure the flag is added specifically to the test_api target, isn't it?

The current implementation is good enough by ensures that:
- The flag is only added for GCC 12 or higher
- The flag is only added if it's supported by the compiler
- The flag is added specifically to the test_api target, not globally

Correct me if I am wrong.

Reply:
Hi @cyyever 
Ack Done.
Please review 
(cc @malfet )